modyn/trainer_server/internal/trainer/remote_downsamplers/remote_uncertainty_downsampling_strategy.py

```diff
@@ -69,28 +69,42 @@ def inform_samples(
         self.index_sampleid_map += sample_ids
 
     def _compute_score(self, forward_output: torch.Tensor, disable_softmax: bool = False) -> np.ndarray:
+        if forward_output.dim() == 1:
+            forward_output = forward_output.unsqueeze(1)
+        feature_size = forward_output.size(1)
+
         if self.score_metric == "LeastConfidence":
-            scores = forward_output.max(dim=1).values.cpu().numpy()
-        elif self.score_metric == "Entropy":
-            preds = (
-                torch.nn.functional.softmax(forward_output, dim=1).cpu().numpy()
-                if not disable_softmax
-                else forward_output.cpu().numpy()
-            )
-            scores = (np.log(preds + 1e-6) * preds).sum(axis=1)
-        elif self.score_metric == "Margin":
-            preds = torch.nn.functional.softmax(forward_output, dim=1) if not disable_softmax else forward_output
-            preds_argmax = torch.argmax(preds, dim=1)  # gets top class
-            max_preds = preds[torch.ones(preds.shape[0], dtype=bool), preds_argmax].clone()  # gets scores of top class
-
-            # remove highest class from softmax output
-            preds[torch.ones(preds.shape[0], dtype=bool), preds_argmax] = -1.0
-
-            preds_sub_argmax = torch.argmax(preds, dim=1)  # gets new top class (=> 2nd top class)
-            second_max_preds = preds[torch.ones(preds.shape[0], dtype=bool), preds_sub_argmax]
-            scores = (max_preds - second_max_preds).cpu().numpy()
+            if feature_size == 1:
+                # for binary classification comparing how far away the element is from 0.5 after sigmoid layer
+                # is the same as comparing the absolute value of the element before sigmoid layer
```

Comment:
So for least confidence we score each sample by the class that has the highest output value. So for each sample we first need to find which class has the highest output value.

For binary classification we have two classes, so we need to *compare* their output value.

Comparing their output value *after sigmoid layer* is equivalent to comparing the output value of any one class of the two, *after sigmoid layer*, to `0.5` as their sum is one.

Furthermore, comparing the output value of any one class of the two *after sigmoid layer* to `0.5` is equivalent to comparing the output value of any one class of the two *before sigmoid layer* to `0`. The larger (and hence the positive) one has the value as the absolute value of any output value.

We assume the model output is not sigmoid transformed, because we use the `BCEWithLogitsLoss` loss as the loss function. This loss function contains a sigmoid layer. Also this is similar to the assumption that for multi-classification we assume the model output is not softmax transformed, as `CrossEntropyLoss` has the softmax layer.

Reply:
Does this answer your question? If so, I will make the comments more detailed for easy comprehension.

Reply:
Sorry, I think I am still either missing something or am confused. `sigmoid` maps any arbitrary real to (0,1), i.e., the input is 1D.

"Comparing their output value after sigmoid layer is equivalent to comparing the output value of any one class of the two, after sigmoid layer, to 0.5 as their sum is one."

I do not understand this. I don't understand what are you comparing. The input to sigmoid is 1D, and the output is 1D, but you're transforming it to a probability. What I do agree with if the input grows/shrinks, then the sigmoid probability also grows/shrinks. So for f(x) = sigmoid(x), if x is very small, the model will be very certain (that it is class 0), if x is very large, the model will be very certain (that it is class 1), and only for x inbetween it will be uncertain (probability around 0.5). 

In the 2D case, we take the lowest probability as the score in the vector. What I would've expected here is that you transform the output like you also do for margin/entropy using actually sigmoid and torch.cat. It might be that we can avoid this and this is cheaper, but I do still not understand where torch.abs is coming from. Because both large and small x can imply uncertainty/certainty, if I am not mistaken. A very low probability also means the model is certain? But I might totally be missing something, sorry

Reply:
Ok I only now realized that sigmoid(0) = 0.5, and hence your abs approach indeed works! Maybe you could add a comment that a model output of 0 represents total uncertainty here, and the higher the absolute value the higher/lower the probability, hence this works.

Reply:
I added detailed explanation in the comment. Let me know if we can merge this PR. Thanks!